pageserver/src/tenant/timeline/timeline_wrappers.rs

```diff
@@ -0,0 +1,169 @@
+//
+// `TimelineGuard` would need more improvements later, in particular, we need to track such guards and wait for them to stop before
+// graceful shutdown.
+impl TimelineRef {
+    // Let only Timeline module the ability to create weak references on itself.
+    pub(super) fn new(timeline: &Timeline) -> Self {
+        Self {
+            timeline: Weak::clone(&timeline.myself),
+            id: TenantTimelineId::new(timeline.tenant_id, timeline.timeline_id),
+        }
+    }
+
+    /// Successfully upgrades to [`TimelineGuard`] if the timeline is in pageserver's memory.
+    /// No state checks are made, only timeline liveness itself.
+    pub fn timeline(&self) -> anyhow::Result<TimelineGuard> {
+        Ok(TimelineGuard {
+            timeline: self
+                .timeline
+                .upgrade()
+                .with_context(|| format!("Timeline {} is dropped", self.id))?,
+        })
+    }
+
+    /// Successfully upgrades to [`TimelineGuard`] if the timeline is in pageserver's memory and
+    /// has [`TimelineState::Active`] state when upgraded.
+    ///
+    /// Further state checks are user's responsibility, the active state can change.
+    pub fn active_timeline(&self) -> anyhow::Result<TimelineGuard> {
+        let timeline = self.timeline()?;
+        anyhow::ensure!(
+            timeline.is_active(),
+            "Timeline {} is not active, state: {:?}",
+            self.id,
+            timeline.current_state(),
+        );
+        Ok(timeline)
+    }
+
+    /// Attempts to upgrade the timeline ref and wait for the timeline to get [`TimelineState::Active`] state (if not already).
+    /// Bails if the timeout expires or timeline fails to become active (e.g. gets into end state that never transitions to active).
+    ///
+    /// Further state checks are user's responsibility, the active state can change.
+    pub async fn get_active_timeline_with_timeout(&self) -> anyhow::Result<TimelineGuard> {
+        let timeline = self.timeline()?;
+        match tokio::time::timeout(
+            ACTIVE_TIMELINE_WAIT_TIMEOUT,
+            timeline.wait_to_become_active(),
+        )
+        .await
+        {
+            Ok(Ok(())) => Ok(timeline),
+            Ok(Err(e)) => Err(e),
+            Err(_) => anyhow::bail!("Timeout waiting for timeline {} to become Active", self.id),
+        }
+    }
+}
+
+/// A wrapper around `Arc<Timeline>` that prevents it from being dropped fully and allows accessing timeline's methods.
+/// Not intended to be cloned or stored somewhere: use it to access timeline's properties and drop it afterwards.
+/// By reacquiring the guard periodically, the timeline-accessing processes can stop when the timeline is being shut down.
+pub struct TimelineGuard {
+    timeline: Arc<Timeline>,
+}
+
+// All `Timeline::` methods here exist because `Arc<Timeline>` is needed for them to run
```

Comment:
You could change all these functions to take `&self` as the argument, and call `self.myself.upgrade()` in the function

Reply:
That's interesting, a bit less clear now, which methods need such clones, but done.